### ISSUE TITLE
Clarify relationship between accelerator mapping, device object, queue ...

### ISSUE THREAD

**Issue by SimeonEhrig** [alpaka-group/alpaka#1581]
I want to overwork the API of vikunja (https://github.com/alpaka-group/vikunja/issues/44). Therefore I think about, which template types and function arguments are necessary. For example, do I need to pass the device object and queue object, or can I get the device from the queue. At the moment, the documentation about this relationship is missing.

**Comment by fwyzard** [comment]
You can get the device type from the queue type:
```c++
using Device = alpaka::Dev<Queue>;
```
and the device object from the queue object:
```c++
Device device = alpaka::getDev(queue);
```

**Comment by SimeonEhrig** [comment]
> You can get the device type from the queue type:
> 
> ```c++
> using Device = alpaka::Dev<Queue>;
> ```
> 
> and the device object from the queue object:
> 
> ```c++
> Device device = alpaka::getDev(queue);
> ```

Thanks for the hint. But I thought @j-stephan mentioned a way, how to create a queue without a device, in the VC? It should be used in the test cases.

**Comment by j-stephan** [comment]
Yes, this is possible for the CPU device:

```c++
using Host = alpaka::DevCpu;
using HostQueue = alpaka::Queue<Host, alpaka::Blocking>;

// CPU host
auto const host = alpaka::getDevByIdx<Host>(0u);
INFO("using alpaka device: " << alpaka::getName(host))
HostQueue hostQueue(host);
```

Live-action example: https://github.com/alpaka-group/alpaka/blob/develop/test/integ/hostOnlyAPI/src/hostOnlyAPI.cpp, line 56 and following.

**Comment by SimeonEhrig** [comment]
> Yes, this is possible for the CPU device:
> 
> ```c++
> using Host = alpaka::DevCpu;
> using HostQueue = alpaka::Queue<Host, alpaka::Blocking>;
> 
> // CPU host
> auto const host = alpaka::getDevByIdx<Host>(0u);
> INFO("using alpaka device: " << alpaka::getName(host))
> HostQueue hostQueue(host);
> ```
> 
> Live-action example: https://github.com/alpaka-group/alpaka/blob/develop/test/integ/hostOnlyAPI/src/hostOnlyAPI.cpp, line 56 and following.

But this looks like a device: `auto const host = alpaka::getDevByIdx<Host>(0u);`

**Comment by j-stephan** [comment]
Ah, I didn't read your question properly. No, it isn't possible to create a queue without a device (what would it be connected to?), only the other way around (device without queue).

**Comment by SimeonEhrig** [comment]
Yes, that makes sense, for example to allocate memory.

**Comment by SimeonEhrig** [comment]
We clarified the relationship today in the VC.

- Each device object has a device type (e.g.: `alpaka::DevCPU`, `alpaka::DevUniformCudaHipRt`)
- Each queue has a ~~acc~~ device object
- Many queues can use a single ~~acc~~ device object
- Each acceleration type (e.g. `alpaka::AccCpuSerial`, `alpaka::AccGpuHipRt`) is dedicated to a device type
  - this means for example, a device object which has the type `alpaka::DevCPU` can be used together with the accelerator type, which starts with `AccCpu`

```c++
auto const devAcc = alpaka::getDevByIdx<alpaka::AccCpuThreads>(0u); // AccCpuThreads is dedicated to alpaka::DevCPU, therefore the device object has the type `alpaka::DevCPU`

// ...

Queue queue(acc);

alpaka::exec<alpaka::AccCpuOmp2Threads>(queue, /* ... */); // is allowed
```

**Comment by fwyzard** [comment]
Mhm, this is not what the code says.

A `queue` has a _device_, not an _accelerator_:
```
namespace alpaka
{
    using QueueCpuBlocking = QueueGenericThreadsBlocking<DevCpu>;
}
```

In principle one could use the same queue with different accelerators on the same device; _e.g._ the serial CPU backend and the TBB backend, or a 1D accelerator and a 3D accelerator, _etc._

**Comment by fwyzard** [comment]
I would say that
  * Each device object has a device type (_e.g._ `alpaka::DevCPU`, `alpaka::DevUniformCudaHipRt`, ...)
  * Each queue object has a queue type (_e.g._ `alpaka::QueueCpuBlocking`, `alpaka::QueueUniformCudaHipRtNonBlocking`, ...) with an underlying device type (see above) and a sync/async implementation
  * Each queue object has an underlying device object
  * Each accelerator object has an accelerator type (_e.g._ `alpaka::AccCpuSerial`, `alpaka::AccCpuTbbBlocks`, `alpaka::AccGpuUniformCudaHipRt`, ...), with an underlying device type (see above) and a dimensionality (`1`, `2`, `3`, ...)
  * An accelerator object _does not_ have an underlying device object
  * Work submitted to an accelerator using a queue determines the device object being used

**Comment by bussmann** [comment]
I think this discussion is a good hint to update the doc.

**Comment by SimeonEhrig** [comment]
> Mhm, this is not what the code says.
> 
> A `queue` has a _device_, not an _accelerator_:
> 
> ```
> namespace alpaka
> {
>     using QueueCpuBlocking = QueueGenericThreadsBlocking<DevCpu>;
> }
> ```
> 
> In principle one could use the same queue with different accelerators on the same device; _e.g._ the serial CPU backend and the TBB backend, or a 1D accelerator and a 3D accelerator, _etc._

You are right. It was a mistake of mine. I constantly switch device and accelerator object.


> * An accelerator object _does not_ have an underlying device object

I'm not sure, if there is accelerator object (maybe you can create one, but it will not used anywhere).

**Comment by j-stephan** [comment]
> I'm not sure, if there is accelerator object (maybe you can create one, but it will not used anywhere).

Accelerator objects cannot be created by the user. They only exist inside a kernel and are created in the background by the respective back-end.

**Comment by SimeonEhrig** [comment]
You are right. I forgot, that the acc objects are there.
